cf-testd/cf-testd.c

```diff
@@ -0,0 +1,492 @@
+    const int encrypted = 0;
+
+    const int received = ReceiveTransaction(conn->conn_info,
+                                            recvbuffer, NULL);
+
+    if (received == -1)
+    {
+        /* Already Log()ged in case of error. */
+        return false;
+    }
+    if (received > CF_BUFSIZE - 1)
+    {
+        UnexpectedError("Received transaction of size %d", received);
+        return false;
+    }
+
+    if (strlen(recvbuffer) == 0)
+    {
+        Log(LOG_LEVEL_WARNING,
+            "Got NULL transmission (of size %d)", received);
+        return true;
+    }
+    /* Don't process request if we're signalled to exit. */
+    if (IsPendingTermination())
+    {
+        Log(LOG_LEVEL_VERBOSE, "Server must exit, closing connection");
+        return false;
+    }
+
+    // See BusyWithNewProtocol() in server_tls.c for how to add other commands
+    switch (GetCommandNew(recvbuffer))
+    {
+    case PROTOCOL_COMMAND_QUERY:
+    {
+        char query[256], name[128];
+        int ret1 = sscanf(recvbuffer, "QUERY %255[^\n]", query);
+        int ret2 = sscanf(recvbuffer, "QUERY %127s", name);
+        if (ret1 != 1 || ret2 != 1)
+        {
+            return CFTestD_ProtocolError(conn, recvbuffer, sendbuffer);
+        }
+
+        if (CFTestD_GetServerQuery(conn, recvbuffer, encrypted))
+        {
+            return true;
+        }
+
+        break;
+    }
+    case PROTOCOL_COMMAND_BAD:
+    default:
+        Log(LOG_LEVEL_WARNING, "Unexpected protocol command: %s", recvbuffer);
+    }
+
+    return CFTestD_ProtocolError(conn, recvbuffer, sendbuffer);
+}
+
+static ServerConnectionState *CFTestD_NewConn(ConnectionInfo *info)
+{
+#if 1
+    /* TODO: why do we do this ?  We fail if getsockname() fails, but
+     * don't use the information it returned.  Was the intent to use
+     * it to fill in conn->ipaddr ? */
+    struct sockaddr_storage addr;
+    socklen_t size = sizeof(addr);
+
+    if (getsockname(ConnectionInfoSocket(info), (struct sockaddr *)&addr, &size) == -1)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Could not obtain socket address. (getsockname: '%s')",
+            GetErrorStr());
+        return NULL;
+    }
+#endif
+
+    ServerConnectionState *conn = xcalloc(1, sizeof(*conn));
+    conn->ctx = NULL;
+    conn->conn_info = info;
+    conn->encryption_type = 'c';
+    /* Only public files (chmod o+r) accessible to non-root */
+    conn->uid = CF_UNKNOWN_OWNER;                    /* Careful, 0 is root! */
+    /* conn->maproot is false: only public files (chmod o+r) are accessible */
+
+    Log(LOG_LEVEL_DEBUG,
+        "New connection (socket %d).",
+        ConnectionInfoSocket(info));
+    return conn;
+}
+
+static void CFTestD_DeleteConn(ServerConnectionState *conn)
+{
+    int sd = ConnectionInfoSocket(conn->conn_info);
+    if (sd != SOCKET_INVALID)
+    {
+        cf_closesocket(sd);
+    }
+    ConnectionInfoDestroy(&conn->conn_info);
+
+    *conn = (ServerConnectionState) {0};
```

Comment:
What is this line about?

Reply:
Don't know. Copied from cf-serverd. Maybe @jimis can elaborate?

Reply:
zeroing the memory?

Reply:
Could be, but the zeroed memory is used (the pointer is dereferenced again) on the line below this.